light-prover/merkle-tree/indexed_merkle_tree.go

```diff
@@ -0,0 +1,329 @@
+	CurrentNodeIndex uint32
+	HighestNodeIndex uint32
+}
+
+type IndexedElement struct {
+	Value     *big.Int
+	NextValue *big.Int
+	NextIndex uint32
+	Index     uint32
+}
+
+type IndexedElementBundle struct {
+	NewLowElement       IndexedElement
+	NewElement          IndexedElement
+	NewElementNextValue *big.Int
+}
+
+type IndexedMerkleTree struct {
+	Tree       *PoseidonTree
+	IndexArray *IndexedArray
+}
+
+func NewIndexedMerkleTree(height uint32) (*IndexedMerkleTree, error) {
+	tree := NewTree(int(height))
+	indexArray := &IndexedArray{
+		Elements: []IndexedElement{{
+			Value:     big.NewInt(0),
+			NextValue: big.NewInt(0),
+			NextIndex: 0,
+			Index:     0,
+		}},
+		CurrentNodeIndex: 0,
+		HighestNodeIndex: 0,
+	}
+
+	return &IndexedMerkleTree{
+		Tree:       &tree,
+		IndexArray: indexArray,
+	}, nil
+}
+
+func (ia *IndexedArray) Init() error {
+	maxAddr := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 248), big.NewInt(1))
+
+	bundle := IndexedElementBundle{
+		NewLowElement: IndexedElement{
+			Value:     big.NewInt(0),
+			NextValue: maxAddr,
+			NextIndex: 1,
+			Index:     0,
+		},
+		NewElement: IndexedElement{
+			Value:     maxAddr,
+			NextValue: big.NewInt(0),
+			NextIndex: 0,
+			Index:     1,
+		},
+		NewElementNextValue: big.NewInt(0),
+	}
+
+	ia.Elements = []IndexedElement{bundle.NewLowElement, bundle.NewElement}
+	ia.CurrentNodeIndex = 1
+	ia.HighestNodeIndex = 1
+
+	return nil
+}
+
+func (ia *IndexedArray) Get(index uint32) *IndexedElement {
+	if int(index) >= len(ia.Elements) {
+		return nil
+	}
+	return &ia.Elements[index]
+}
+
+func (ia *IndexedArray) Append(value *big.Int) error {
+	lowElementIndex := ia.FindLowElementIndex(value)
+	lowElement := ia.Elements[lowElementIndex]
+
+	if lowElement.NextIndex != 0 {
+		nextElement := ia.Elements[lowElement.NextIndex]
+		if value.Cmp(nextElement.Value) >= 0 {
+			return fmt.Errorf("new value must be less than next element value")
+		}
+	}
+
+	newElementIndex := uint32(len(ia.Elements))
+	newElement := IndexedElement{
+		Value:     value,
+		NextValue: lowElement.NextValue,
+		NextIndex: lowElement.NextIndex,
+		Index:     newElementIndex,
+	}
+
+	ia.Elements[lowElementIndex].NextIndex = newElementIndex
+	ia.Elements[lowElementIndex].NextValue = value
+
+	ia.Elements = append(ia.Elements, newElement)
+	ia.CurrentNodeIndex = newElementIndex
+	if lowElement.NextIndex == 0 {
+		ia.HighestNodeIndex = newElementIndex
+	}
+
+	return nil
+}
+func (ia *IndexedArray) FindLowElementIndex(value *big.Int) uint32 {
+	maxAddr := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 248), big.NewInt(1))
+
+	// If we only have initial elements (0 and maxAddr)
+	if len(ia.Elements) == 2 {
+		// Always return the first element (0) as low element
+		return 0
+	}
+
+	for i, element := range ia.Elements {
+		// Skip the max element
+		if element.Value.Cmp(maxAddr) == 0 {
+			continue
+		}
+
+		// If this is the last element in chain
+		if element.NextIndex == 0 {
+			return uint32(i)
+		}
```

Comment:
```suggestion

```
Since we iterate over the elements in the order of the indexed array but not in ascending order by value the fact that we found the element highest element doesn't mean that we should stop looking for a range.